sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -29,7 +30,19 @@ from ._k_means_fast cimport _center_shift
 np.import_array()
 
 
-def _init_bounds_dense(
+# Threadpoolctl wrappers to limit the number of threads in second level of
+# nested parallelism (i.e. BLAS) to avoid oversubsciption.
+def elkan_iter_chunked_dense(*args, **kwargs):
+    with threadpool_limits(limits=1, user_api="blas"):
```

Comment:
Thoughts regarding using the CM that deep in the code, versus doing it sooner e.g. in `_kmeans_single_lloyd` after `_init_centroids` is called?

I guess we may miss oversubscription issues if one day we decide to re-write `_inertia_*`

Only curious what you think

Reply:
> I guess we may miss oversubscription issues if one day we decide to re-write _inertia_*

True, but right now it uses no prange and no blas so I don't see it changing soon :)
And more importantly, we should think of the implications each time we want to add a prange somewhere.

> Thoughts regarding using the CM that deep in the code, versus doing it sooner e.g. in _kmeans_single_lloyd after _init_centroids is called?

I wanted to put it where we're sure there will be no BLAS call took inside the context manager. That way we make sure that we can change the python code more safely.

Besides,  In _kmeans_single_elkan you have:
```py
for i in range(max_iter):
    elkan_iter(...)
    ...
    euclidean_distances(...)  # BLAS
```
we don't want to include euclidean distances in the CM. So we have to put the CM around elkan_iter.